In [3]:
import numpy as np


# 1. Datos del problema
t = np.array([0, 1, 3, 5, 7, 9], dtype=float)
gamma = np.array([1.0, 0.891, 0.708, 0.562, 0.447, 0.355], dtype=float)

# 2. Función no lineal f(lambda)
def f(lmbda):
    term1 = np.sum(gamma * t * np.exp(lmbda * t))
    term2 = (
        np.sum(gamma * np.exp(lmbda * t)) /
        np.sum(np.exp(2 * lmbda * t))
    ) * np.sum(t * np.exp(2 * lmbda * t))
    return term1 - term2

# 3. M. bisección

def biseccion(func, xl, xu, tol=1e-8, max_iter=1000):
    if func(xl) * func(xu) > 0:
        raise ValueError("El intervalo no encierra una raíz.")

    iteraciones = 0

    while abs(xu - xl) > tol and iteraciones < max_iter:
        xm = (xl + xu) / 2

        if func(xl) * func(xm) < 0:
            xu = xm
        else:
            xl = xm

        iteraciones += 1

    raiz = (xl + xu) / 2
    return raiz, iteraciones

# 4. Límites de bisección con nc = 49 (mis 2 ultimos digitos de la cedula)
lambda_aprox = -0.11505
nc = 49

delta = nc / 1000

xl = lambda_aprox - delta
xu = lambda_aprox + delta

print("Intervalo de bisección:")
print(f"xl = {xl:.8f}")
print(f"xu = {xu:.8f}")


# 5.Lambda

lambda_sol, n_iter = biseccion(f, xl, xu, tol=1e-8)

print("\nResultado de bisección:")
print(f"lambda = {lambda_sol:.8f}")
print(f"iteraciones = {n_iter}")


# 6. A
A = np.sum(gamma * np.exp(lambda_sol * t)) / np.sum(np.exp(2 * lambda_sol * t))

print(f"\nA = {A:.8f}")


# 7. Vida media
t_half = np.log(2) / abs(lambda_sol)

print(f"\nVida media = {t_half:.8f} horas")


# 8. Intensidad después de 24 horas
gamma_24 = A * np.exp(lambda_sol * 24)

print(f"\nIntensidad después de 24 h = {gamma_24:.8f}")

Intervalo de bisección:
xl = -0.16405000
xu = -0.06605000

Resultado de bisección:
lambda = -0.11508304
iteraciones = 24

A = 0.99984140

Vida media = 6.02301759 horas

Intensidad después de 24 h = 0.06315574


<div style="background-color:#ffe6f0; padding:18px; border-radius:12px; border:1.5px solid #ffb3d1;">

### Determinación de los parámetros $A$ y $\lambda$

Para describir la desintegración radiactiva se usa el modelo exponencial:

$$
\gamma(t)=Ae^{\lambda t}
$$

Para encontrar el valor de $\lambda$ se construye la función no lineal dada en el enunciado:

$$
f(\lambda)=\sum_{i=1}^{n}\gamma_i t_i e^{\lambda t_i}
-
\frac{\sum_{i=1}^{n}\gamma_i e^{\lambda t_i}}{\sum_{i=1}^{n}e^{2\lambda t_i}}
\sum_{i=1}^{n} t_i e^{2\lambda t_i}
$$

La raíz de esta función se obtiene numéricamente usando el **método de bisección**.

Del ejemplo anterior se toma el valor aproximado $\lambda \approx -0.11505$.  
Como los dos últimos dígitos de mi cédula son **49**, el intervalo inicial se construye como

$$
\lambda \pm 0.049
$$

por lo que el intervalo de bisección utilizado es

$$
[x_l,x_u]=[-0.16405,-0.06605]
$$

El algoritmo se ejecuta con una tolerancia de

$$
10^{-8}
$$

Aplicando el método se obtiene:

- $\lambda \approx -0.11508304$
- $A \approx 0.99984140$
- vida media $t_{1/2}\approx 6.02$ horas
- $\gamma(24)\approx 0.06316$

</div>